### ISSUE TITLE
Archiving removes unexpected files

### ISSUE THREAD

**Issue by rymk** [NLog/NLog#1681]
**Bug**
**NLog version**: 4.3.3 & 4.3.8  
**Platform**: .NET 4.5.2

**Scenario:**
We have multiple .NET apps running on the same server. Each app writes to its own log file, but all logs are in the same directory (using naming conventions to distinguish between apps). Similarly, all the logs are archived to a different directory but all apps share the same archive directory. _See the example configs below to see the hierarchy_

**Example configs:**
_**App1:**_

``` xml
<nlog xmlns:nlog="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <targets>
      <!-- Log in a separate thread, possibly queueing up to 5000 messages. When the queue overflows, discard any extra messages-->
      <default-wrapper xsi:type="AsyncWrapper" queueLimit="5000" overflowAction="Discard">
      </default-wrapper>
      <target name="logfile" xsi:type="File" fileName="C:\Logs\Project.Web.log"
                archiveFileName="C:\Logs\Archives\${shortdate}-Project.Web.log"
                archiveEvery="Day" maxArchiveFiles="7"
                layout="${longdate} [${level}] [${callsite}] ${message}"
                concurrentWrites="true" keepFileOpen="false" />
    </targets>
    <rules>
      <logger name="*" minLevel="Debug" writeTo="logfile" />
    </rules>
</nlog>
```

_**App2:**_

``` xml
<nlog xmlns:nlog="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <targets>
      <!-- Log in a separate thread, possibly queueing up to 5000 messages. When the queue overflows, discard any extra messages-->
      <default-wrapper xsi:type="AsyncWrapper" queueLimit="5000" overflowAction="Discard">
      </default-wrapper>
      <target name="logfile" xsi:type="File" fileName="C:\Logs\Assembly.Web.log"
                archiveFileName="C:\Logs\Archives\${shortdate}-Assembly.Web.log"
                archiveEvery="Day" maxArchiveFiles="7"
                layout="${longdate} [${level}] [${callsite}] ${message}"
                concurrentWrites="true" keepFileOpen="false" />
    </targets>
    <rules>
      <logger name="*" minLevel="Debug" writeTo="logfile" />
    </rules>
</nlog>
```

**Expected Result**
Both apps to be kept for 7 days. 

**What is actually happening**
Note: We have more than 7 apps logging and archiving to the same directory on the server in question. When we look at the archive directory each day, we are only seeing the archived log file from the previous day. It almost looks like NLog is deciding to keep 7 log files in the Archive directory, regardless of if they are all ${shortdate}-Assembly.Web.log vs ${shortdate}-Project.Web.log.

Not all of the apps on this server have been upgrade to NLog 4.3.x+, but even the apps with the older version of NLog are having their archives deleted prematurely; assuming because 1 of the apps with version 4.3.x is deleting them too.

**Previously Worked**
Last known to work on 4.0.1 (and earlier on some apps)

Seems somewhat related to [Issue 1670](https://github.com/NLog/NLog/issues/1670)

I assume if we wrote the logs and archives to their own unique directories per application we would not have this issue, but we do not have time allocated to make these changes in all apps right now.

Thank you for your assistance.

**Comment by 304NotModified** [comment]
Could you enable the internal log (trace) and send me that one when it occurs? Thanks!

I also think this bug has been introduced in 4.2. If you could confirm that, that would be great

**Comment by rymk** [comment]
I was having issues uploading the zip, below is a temporary link to the zip.
https://qorvo-my.sharepoint.com/personal/ryan_key_qorvo_com/_layouts/15/guestaccess.aspx?guestaccesstoken=aJiQczLbrzY7JDzg%2f6P10GwlsoKZ2mWyCeRi4MJXlvY%3d&docid=1f7d14bddb7f8405a88807eb4961a9159&rev=1&expiration=2016-11-06T00%3a01%3a03.000Z 

I also found that we have had an application running without causing problems using NLOG 4.2.3.

**Comment by 304NotModified** [comment]
> also found that we have had an application running without causing problems using NLog 4.2.3.

OK thanks!

**Comment by 304NotModified** [comment]
Could you help writing an unit test? See for example; https://github.com/NLog/NLog/pull/1680/files#diff-8b2a0fab8f5b0594b3db553894939a44

That would really help & boost the bug fix.

**Comment by rymk** [comment]
Sure, I can work on this today. I looked over the code yesterday; seems fairly straightforward. 

It looks like the unit test should be a similar variation to the example you provided and https://github.com/NLog/NLog/blob/493c5c1d9a9e70d5c057491fcd4d38dd07b9b6a8/tests/NLog.UnitTests/Targets/FileTargetTests.cs#L2558-L2633 

I'll let you know if I run into any unexpected roadblocks.

**Comment by 304NotModified** [comment]
Cool! Thanks in advance! 

If it's possible to fix the unit this before Tuesday, I will try to fix it for 4.3.10. Otherwise I will be probable 4.3.11

**Comment by 304NotModified** [comment]
Working on this, but are you sure that this worked before? Some code is then really missing.

**Comment by 304NotModified** [comment]
Tested this with FileTarget from NLog 4.0.1 and this is also the issue there. 

When reading the files in the archiveFolder, there isn't a filemask. If it was implemented, then it should be constructed from the archiveFileName. E.g. `C:\Logs\Archives\${shortdate}-Assembly.Web.log` should be `C:\Logs\Archives\*-Assembly.Web.log`

To bad there is only code that is doing this: (when using date&sequence or only sequencce) `C:\Logs\Archives\${shortdate}-{#}-Assembly.Web.log` => `C:\Logs\Archives\${shortdate}-*-Assembly.Web.log`.

**Comment by rymk** [comment]
To answer if it was working before, it did not appear to be working at all before.
Most of our apps have/had configs similar to (each with different file names)
```
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <targets>
      <!-- Log in a separate thread, possibly queueing up to 5000 messages. When the queue overflows, discard any extra messages-->
      <target name="logfile" xsi:type="AsyncWrapper" queueLimit="5000" overflowAction="Discard">
        <target xsi:type="File" fileName="C:\Logs\Qorvo.ProjectCenter.Web.log" archiveFileName="C:\Logs\Archives\${shortdate}-Qorvo.ProjectCenter.Web.log" archiveEvery="Day" maxArchiveFiles="7" layout="${longdate} [${level}] [${callsite}] ${message}" concurrentWrites="true" keepFileOpen="false" />
      </target>
    </targets>
    <rules>
      <logger name="*" minLevel="Debug" writeTo="logfile" />
    </rules>
  </nlog>
```

Interestingly, even though these are all set to maxArchiveFiles=7, the archive files were NEVER deleted. Part of our architecture includes a scheduled job that is run daily; that job would then go into the archive directory and delete any file older than 90 days (based on last modified date I believe). So we were keeping 90 days of history per application since NLog is configured to archvie daily.

Are you proposing that if we change our archive file naming convention to include the ${shortdate}-**{#}**- then our extra files should stop getting deleted? 

It is interesting, because even though we aren't defining {#} in the archiveFileName, the logs are still written as _C:\Logs\Archives\${shortdate}-Assembly.Web**-{#}**.log_ in the 4.3.x+ version of NLog.

**Comment by 304NotModified** [comment]
This is an issue, bug it's not trivial to fix.

**Comment by 304NotModified** [comment]
Closing this as this a duplicate of https://github.com/NLog/NLog/issues/1670


NB: we really like to fix this one! Any help would be appreciated

**Comment by snakefoot** [comment]
@rymk If you use **archiveNumbering="Date"** and replaced **{shortdate}** with **{#}**, then your archive cleanup should work properly.

**Comment by 304NotModified** [comment]
fixed by #1993, NLog 4.5. Thanks @snakefoot !
